fontbe/src/hvar.rs

```diff
@@ -0,0 +1,233 @@
+            // this glyph defines its own set of locations, a new sparse model is needed
+            VariationModel::new(advance_widths.keys().cloned().collect(), self.axes.clone())
+                .unwrap()
+        });
+        self.deltas.push(
+            model
+                .deltas(&advance_widths)
+                .map_err(|e| Error::GlyphDeltaError(name.clone(), e))?
+                .into_iter()
+                .filter_map(|(region, values)| {
+                    if region.is_default() {
+                        return None;
+                    }
+                    // Only 1 value per region for our input
+                    assert!(values.len() == 1, "{} values?!", values.len());
+                    Some((
+                        region.to_write_fonts_variation_region(&self.axes),
+                        values[0].ot_round(),
+                    ))
+                })
+                .collect(),
+        );
+        Ok(())
+    }
+
+    fn is_single_model(&self) -> bool {
+        self.models.len() == 1
+    }
+
+    fn iter(&self) -> impl Iterator<Item = &Vec<(VariationRegion, i16)>> {
+        self.deltas.iter()
+    }
+}
+
+impl Work<Context, AnyWorkId, Error> for HvarWork {
+    fn id(&self) -> AnyWorkId {
+        WorkId::Hvar.into()
+    }
+
+    fn read_access(&self) -> Access<AnyWorkId> {
+        Access::Custom(Arc::new(|id| {
+            matches!(
+                id,
+                AnyWorkId::Fe(FeWorkId::Glyph(..))
+                    | AnyWorkId::Fe(FeWorkId::StaticMetadata)
+                    | AnyWorkId::Fe(FeWorkId::GlyphOrder)
+            )
+        }))
+    }
+
+    /// Generate [HVAR](https://learn.microsoft.com/en-us/typography/opentype/spec/HVAR)
+    fn exec(&self, context: &Context) -> Result<(), Error> {
+        let static_metadata = context.ir.static_metadata.get();
+        let var_model = &static_metadata.variation_model;
+        let glyph_order = context.ir.glyph_order.get();
+
+        let mut glyph_width_deltas = AdvanceWidthDeltas::new(var_model.clone());
+        for name in glyph_order.iter() {
+            let glyph = context.ir.glyphs.get(&FeWorkId::Glyph(name.clone()));
+            glyph_width_deltas.add(glyph.as_ref())?;
+        }
+
+        let mut var_idxes = Vec::new();
+
+        // if we have a single model, we can try to build a VariationStore with implicit variation
+        // indices (a single ItemVariationData, outer index 0, inner index => gid).
+        let direct_store = if glyph_width_deltas.is_single_model() {
+            let mut direct_builder = VariationStoreBuilder::new_with_implicit_indices();
+            for deltas in glyph_width_deltas.iter() {
+                var_idxes.push(direct_builder.add_deltas(deltas.clone()));
+            }
+            // sanity check
+            assert_eq!(
+                mem::take(&mut var_idxes),
+                (0..glyph_order.len() as u32).collect::<Vec<_>>()
+            );
```

Comment:
`mem::take` replaces `var_idxes` with an empty vector, which feels weird. is `var_idxes` only constructed for the purposes of this assert?

In any case, there are likely more efficient ways to write this. One simple one would be to avoid the vector, and instead just do the comparison with a length check followed by direct iteration, e.g.

```rust
assert_eq!(var_idxes.len(), glyph_order.len());
assert!(var_idxes.iter().enumerate().all(|(i, idx)| i as u32 == *idx));
```

Reply:
thanks, I applied most of your suggestions in the cleaned up PR => https://github.com/googlefonts/fontc/pull/527/commits/c724502fb6d34cdff6e7c239b2f8276eda130829